syft/pkg/cataloger/golang/scan_binary.go

```diff
@@ -89,6 +116,277 @@ func getCryptoSettingsFromVersion(v version.Version) []string {
 	return cryptoSettings
 }
 
+// getCachedChecksum just as the function name, it gets the checksum from the path with the pattern like
+// $GOPATH/pkg/mod/cache/download/example.com/module/@v/v1.2.3.ziphash
+func getCachedChecksum(pkgDir fs.FS, name string, version string) (content string, err error) {
+	var res string
+	err = fs.WalkDir(pkgDir, ".", func(filePath string, _ fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("error when walking down the dir to find %s", name)
+		}
+		if strings.HasPrefix(filePath, version+".ziphash") {
+			res, err = fetchFileContents(pkgDir, filePath)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if res == "" {
+		return "", fmt.Errorf("no checksum for %s@%s", name, version)
+	}
+	return res, err
+}
+
+func trimmedAsURL(name string) (urlDir string, urlName string) {
+	if strings.HasPrefix(name, "vendor") {
+		return
+	}
+	parts := strings.Split(name, "/")
+	var lastIndex int
+	if len(parts) >= 4 {
+		verLikely := strings.Split(parts[3], ".")[0]
+		pattern := `^v\d+`
+		re := regexp.MustCompile(pattern)
+		match := re.MatchString(verLikely)
+		lastIndex = 2
+		if match {
+			lastIndex = 3
+		}
+	} else {
+		lastIndex = len(parts) - 1
+	}
+	versionWithFunc := parts[lastIndex]
+	if idx := strings.Index(versionWithFunc, "."); idx != -1 {
+		parts[lastIndex] = versionWithFunc[:idx]
+	}
+	if lastIndex != 0 {
+		urlDir = strings.Join(parts[:lastIndex], "/")
+	}
+	urlName = parts[lastIndex]
+	return
+}
+
+func findVersionInCache(basePath fs.FS, baseName string) (string, error) {
+	var res string
+	err := fs.WalkDir(basePath, ".", func(filePath string, _ fs.DirEntry, err error) error {
```

Comment:
instead of using WalkDir should this just iterate the direct subdirectories?

(same comment for other WalkDir calls)